// @acala-network/chopsticks-core/dist/cjs/rpc/rpc-spec/chainHead_v1.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    chainHead_v1_body: function() {
        return chainHead_v1_body;
    },
    chainHead_v1_call: function() {
        return chainHead_v1_call;
    },
    chainHead_v1_continue: function() {
        return chainHead_v1_continue;
    },
    chainHead_v1_follow: function() {
        return chainHead_v1_follow;
    },
    chainHead_v1_header: function() {
        return chainHead_v1_header;
    },
    chainHead_v1_stopOperation: function() {
        return chainHead_v1_stopOperation;
    },
    chainHead_v1_storage: function() {
        return chainHead_v1_storage;
    },
    chainHead_v1_unfollow: function() {
        return chainHead_v1_unfollow;
    },
    chainHead_v1_unpin: function() {
        return chainHead_v1_unpin;
    }
});
const _logger = require("../../logger.js");
const _shared = require("../shared.js");
const logger = _logger.defaultLogger.child({
    name: 'rpc-chainHead_v1'
});
const following = new Map();
async function afterResponse(fn) {
    await new Promise((resolve)=>setTimeout(resolve, 0));
    fn();
}
const chainHead_v1_follow = async (context, [withRuntime], { subscribe })=>{
    const update = async (block)=>{
        logger.trace({
            hash: block.hash
        }, 'chainHead_v1_follow');
        const getNewRuntime = async ()=>{
            const [runtime, previousRuntime] = await Promise.all([
                block.runtimeVersion,
                block.parentBlock.then((b)=>b?.runtimeVersion)
            ]);
            const hasNewRuntime = runtime.implVersion !== previousRuntime?.implVersion || runtime.specVersion !== previousRuntime.specVersion;
            return hasNewRuntime ? runtime : null;
        };
        const newRuntime = withRuntime ? await getNewRuntime() : null;
        callback({
            event: 'newBlock',
            blockHash: block.hash,
            parentBlockHash: (await block.parentBlock)?.hash,
            newRuntime
        });
        callback({
            event: 'bestBlockChanged',
            bestBlockHash: block.hash
        });
        callback({
            event: 'finalized',
            finalizedBlockHashes: [
                block.hash
            ],
            prunedBlockHashes: []
        });
        const storageDiffs = following.get(id)?.storageDiffs;
        if (storageDiffs?.size) {
            // Fetch the storage diffs and update the `closestDescendantMerkleValue` for those that changed
            const diffKeys = Object.keys(await block.storageDiff());
            for (const [prefix, value] of storageDiffs.entries()){
                if (diffKeys.some((key)=>key.startsWith(prefix))) {
                    storageDiffs.set(prefix, value + 1);
                }
            }
        }
    };
    const id = context.chain.headState.subscribeHead(update);
    const cleanup = ()=>{
        context.chain.headState.unsubscribeHead(id);
        following.delete(id);
    };
    const callback = subscribe('chainHead_v1_followEvent', id, cleanup);
    following.set(id, {
        callback,
        pendingDescendantValues: new Map(),
        storageDiffs: new Map()
    });
    afterResponse(async ()=>{
        callback({
            event: 'initialized',
            finalizedBlockHashes: [
                context.chain.head.hash
            ],
            finalizedBlockRuntime: withRuntime ? await context.chain.head.runtimeVersion : null
        });
    });
    return id;
};
const chainHead_v1_unfollow = async (_, [followSubscription], { unsubscribe })=>{
    unsubscribe(followSubscription);
    return null;
};
const chainHead_v1_header = async (context, [followSubscription, hash])=>{
    if (!following.has(followSubscription)) return null;
    const block = await context.chain.getBlock(hash);
    return block ? (await block.header).toHex() : null;
};
const operationStarted = (operationId)=>({
        result: 'started',
        operationId
    });
const randomId = ()=>Math.random().toString(36).substring(2);
const chainHead_v1_call = async (context, [followSubscription, hash, method, callParameters])=>{
    const operationId = randomId();
    afterResponse(async ()=>{
        const block = await context.chain.getBlock(hash);
        if (!block) {
            following.get(followSubscription)?.callback({
                event: 'operationError',
                operationId,
                error: `Block ${hash} not found`
            });
        } else {
            try {
                const resp = await block.call(method, [
                    callParameters
                ]);
                following.get(followSubscription)?.callback({
                    event: 'operationCallDone',
                    operationId,
                    output: resp.result
                });
            } catch (ex) {
                following.get(followSubscription)?.callback({
                    event: 'operationError',
                    operationId,
                    error: ex.message
                });
            }
        }
    });
    return operationStarted(operationId);
};
const PAGE_SIZE = 1000;
async function getDescendantValues(block, params) {
    const keys = await block.getKeysPaged({
        ...params,
        pageSize: PAGE_SIZE
    });
    const items = await Promise.all(keys.map((key)=>block.get(key).then((value)=>({
                key,
                value
            }))));
    if (keys.length < PAGE_SIZE) {
        return {
            items,
            next: null
        };
    }
    return {
        items,
        next: {
            ...params,
            startKey: keys[PAGE_SIZE - 1]
        }
    };
}
const chainHead_v1_storage = async (context, [followSubscription, hash, items, _childTrie])=>{
    const operationId = randomId();
    afterResponse(async ()=>{
        const block = await context.chain.getBlock(hash);
        if (!block) {
            following.get(followSubscription)?.callback({
                event: 'operationError',
                operationId,
                error: 'Block not found'
            });
            return;
        }
        const handleStorageItemRequest = async (sir)=>{
            switch(sir.type){
                case 'value':
                    {
                        const value = await block.get(sir.key);
                        if (value) {
                            following.get(followSubscription)?.callback({
                                event: 'operationStorageItems',
                                operationId,
                                items: [
                                    {
                                        key: sir.key,
                                        value
                                    }
                                ]
                            });
                        }
                        return null;
                    }
                case 'descendantsValues':
                    {
                        const { items, next } = await getDescendantValues(block, {
                            prefix: sir.key,
                            startKey: '0x'
                        });
                        following.get(followSubscription)?.callback({
                            event: 'operationStorageItems',
                            operationId,
                            items
                        });
                        return next;
                    }
                case 'closestDescendantMerkleValue':
                    {
                        const followingSubscription = following.get(followSubscription);
                        if (!followingSubscription) return null;
                        if (!followingSubscription.storageDiffs.has(sir.key)) {
                            // Set up a diff watch for this key
                            followingSubscription.storageDiffs.set(sir.key, 0);
                        }
                        followingSubscription.callback({
                            event: 'operationStorageItems',
                            operationId,
                            items: [
                                {
                                    key: sir.key,
                                    closestDescendantMerkleValue: String(followingSubscription.storageDiffs.get(sir.key))
                                }
                            ]
                        });
                        return null;
                    }
                default:
                    // TODO
                    console.warn(`Storage type not implemented ${sir.type}`);
                    return null;
            }
        };
        const listResult = await Promise.all(items.map(handleStorageItemRequest));
        const pending = listResult.filter((v)=>v !== null);
        if (!pending.length) {
            following.get(followSubscription)?.callback({
                event: 'operationStorageDone',
                operationId
            });
        } else {
            const follower = following.get(followSubscription);
            if (follower) {
                follower.pendingDescendantValues.set(operationId, {
                    hash,
                    params: pending
                });
                follower.callback({
                    event: 'operationWaitingForContinue',
                    operationId
                });
            }
        }
    });
    return {
        ...operationStarted(operationId),
        discardedItems: 0
    };
};
const limitReached = {
    result: 'limitReached'
};
const chainHead_v1_body = async (context, [followSubscription, hash])=>{
    if (!following.has(followSubscription)) return limitReached;
    const block = await context.chain.getBlock(hash);
    if (!block) {
        throw new _shared.ResponseError(-32801, 'Block not found');
    }
    const operationId = randomId();
    afterResponse(async ()=>{
        const body = await block.extrinsics;
        following.get(followSubscription)?.callback({
            event: 'operationBodyDone',
            operationId,
            value: body
        });
    });
    return operationStarted(operationId);
};
const chainHead_v1_continue = async (context, [followSubscription, operationId])=>{
    const follower = following.get(followSubscription);
    const pendingOp = follower?.pendingDescendantValues.get(operationId);
    if (!pendingOp || !follower) {
        throw new _shared.ResponseError(-32803, "Operation ID doesn't have anything pending");
    }
    const block = await context.chain.getBlock(pendingOp.hash);
    if (!block) {
        throw new _shared.ResponseError(-32801, 'Block not found');
    }
    afterResponse(async ()=>{
        const handlePendingOperation = async (params)=>{
            const { items, next } = await getDescendantValues(block, params);
            follower.callback({
                event: 'operationStorageItems',
                operationId,
                items
            });
            return next;
        };
        const listResult = await Promise.all(pendingOp.params.map(handlePendingOperation));
        const pending = listResult.filter((v)=>v !== null);
        if (!pending.length) {
            follower.pendingDescendantValues.delete(operationId);
            follower.callback({
                event: 'operationStorageDone',
                operationId
            });
        } else {
            follower.pendingDescendantValues.set(operationId, {
                hash: pendingOp.hash,
                params: pending
            });
            follower.callback({
                event: 'operationWaitingForContinue',
                operationId
            });
        }
    });
    return null;
};
const chainHead_v1_stopOperation = async (_context, [followSubscription, operationId])=>{
    following.get(followSubscription)?.pendingDescendantValues.delete(operationId);
    return null;
};
const chainHead_v1_unpin = async (_context, [_followSubscription, _hashOrHashes])=>{
    return null;
};
